Add support for Database/Database name to render StorageLoc in slurmdbd configuration

jdeamicis opened the pull request:
### Description of changes
* Add support for `Scheduling/SlurmSettings/Database/DatabaseName` to be used to configure `StorageLoc` in the slurmdbd configuration include file generated by ParallelCluster.
* Add some unit tests (InSpec and ChefSpec) for the `config_slurm_accounting.rb` recipe.

### Tests
* Added case in unit test for the rendering of the Slurm configuration include files relevant for Slurm Accounting.
* Manually ran the InSpec and ChefSpec added for the `config_slurm_accounting.rb` recipe.

### References
* https://github.com/aws/aws-parallelcluster/pull/5718

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Add ChefSpec test for config_slurm_accounting recipe

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Split creation of Slurm users from Slurm installation

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Cover config_slurm_accounting.rb with InSpec tests

Cover part of the recipe for the configuration of Slurm Accounting
with InSpec tests runnable on docker.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add support for DatabaseName in slurmdbd config rendering

Add logic to use Database/DatabaseName to render StorageLoc in
slurm_parallelcluster_slurmdbd.con if provided. The default value
remains the name of the cluster (opportunely transformed).

Add unit test case to cover creation of slurmdbd conf include file
with the usage of Database/DatabaseName in the input YAML file.

Add Changelog entry.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Indent slurm_parallelcluster_slurmdbd.conf Jinja template

Add trick to have the logic for the renderization of the template
indented without affecting the final output. This improves the
comprehension of the logic embedded in the template (although
slightly visually impacting).

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

EddyMM commented on cookbooks/aws-parallelcluster-slurm/kitchen.slurm-config.yml:
```diff
@@ -191,3 +191,29 @@ suites:
+        scheduler: 'slurm'
+        #slurm:
+        #  user: 'slurm'
+        #  group: 'slurm'
```
These were used for testing? I think they can be removed.

jdeamicis replied:
Addressed.

EddyMM commented on cookbooks/aws-parallelcluster-slurm/recipes/install/slurm_users.rb:
```diff
@@ -0,0 +1,56 @@
+# Cookbook:: aws-parallelcluster-slurm
+# Recipe:: install_slurm
+#
+# Copyright:: 2013-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
NIT: `...-2023`

jdeamicis replied:
Addressed.

EddyMM commented on test/unit/slurm/test_slurm_config_generator.py:
```diff
@@ -102,14 +102,17 @@ def test_generate_slurm_config_files_memory_scheduling(
 )
-def test_generate_slurm_config_files_slurm_accounting(mocker, test_datadir, tmpdir, slurm_accounting):
+def test_generate_slurm_config_files_slurm_accounting(mocker, test_datadir, tmpdir, slurm_accounting, dbname_passed):
+    input_base_name = "sample_input"
```
Instead of building the names programmatically (which requires inspecting various parts of this function), can we pass the files names via Parametrize
`("input_file, output_file", [...])`.

In this case for example, we had to introduce a new if condition in order to test a new parameter.

jdeamicis replied:
Addressed.

EddyMM commented on cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/templates/slurm_parallelcluster_slurmdbd.conf:
```diff
@@ -1,12 +1,17 @@
-StorageUser={{ scaling_config.Database.UserName }}
-StoragePass=dummy
-{% endif %}
+{# Adding comments at the beginning of each line is a trick to indent the template without affecting the output #}
```
I think we generate the configuration files using the `trim_blocks=True` and `lstrip_blocks=True` which should let us use indentation in the template without it appearing in the generated file.

Do the indentations appear in the generated file?

jdeamicis replied:
I had written a nice comment that got lost :/

Long story short, those options only remove the new lines and spaces caused by the presence of the `{# ... #}` Jinja blocks, not any indentation we may want to apply to the rest of the template to make the logic more understandable. Meaning that yes, if I indent lines that do not have Jinja blocks in them, they will remain indented in the rendered output.

I couldn't find a nicer alternative to this.

My impression is that this trick makes the logic embedded in the template more easy to follow, although it becomes slightly uglier to read. I would be in favor of keeping this trick.

jdeamicis committed:
Address code review comments

Simplify unit test for generation of Slurm accounting configuration files.

Remove commented parameters in slurm kitchen test configuration file.

Modify copyright year in one file.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>
